tiledb/api/src/range.rs

```diff
@@ -0,0 +1,841 @@
+
+macro_rules! single_value_range_from {
+    ($($V:ident : $U:ty),+) => {
+        $(
+            impl From<&[$U; 2]> for SingleValueRange {
+                fn from(value: &[$U; 2]) -> SingleValueRange {
+                    SingleValueRange::$V(value[0], value[1])
+                }
+            }
+        )+
+    }
+}
+
+single_value_range_from!(UInt8: u8, UInt16: u16, UInt32: u32, UInt64: u64);
+single_value_range_from!(Int8: i8, Int16: i16, Int32: i32, Int64: i64);
+single_value_range_from!(Float32: f32, Float64: f64);
+
+#[macro_export]
+macro_rules! single_value_range_go {
+    ($expr:expr, $DT:ident, $start:pat, $end:pat, $then:expr) => {
+        match $expr {
+            SingleValueRange::UInt8($start, $end) => {
+                type $DT = u8;
+                $then
+            }
+            SingleValueRange::UInt16($start, $end) => {
+                type $DT = u16;
+                $then
+            }
+            SingleValueRange::UInt32($start, $end) => {
+                type $DT = u32;
+                $then
+            }
+            SingleValueRange::UInt64($start, $end) => {
+                type $DT = u64;
+                $then
+            }
+            SingleValueRange::Int8($start, $end) => {
+                type $DT = i8;
+                $then
+            }
+            SingleValueRange::Int16($start, $end) => {
+                type $DT = i16;
+                $then
+            }
+            SingleValueRange::Int32($start, $end) => {
+                type $DT = i32;
+                $then
+            }
+            SingleValueRange::Int64($start, $end) => {
+                type $DT = i64;
+                $then
+            }
+            SingleValueRange::Float32($start, $end) => {
+                type $DT = f32;
+                $then
+            }
+            SingleValueRange::Float64($start, $end) => {
+                type $DT = f64;
+                $then
+            }
+        }
+    };
+}
+
+#[derive(Clone, Deserialize, Serialize, PartialEq)]
+pub enum MultiValueRange {
+    UInt8(Box<[u8]>, Box<[u8]>),
+    UInt16(Box<[u16]>, Box<[u16]>),
+    UInt32(Box<[u32]>, Box<[u32]>),
+    UInt64(Box<[u64]>, Box<[u64]>),
+    Int8(Box<[i8]>, Box<[i8]>),
+    Int16(Box<[i16]>, Box<[i16]>),
+    Int32(Box<[i32]>, Box<[i32]>),
+    Int64(Box<[i64]>, Box<[i64]>),
+    Float32(Box<[f32]>, Box<[f32]>),
+    Float64(Box<[f64]>, Box<[f64]>),
+}
+
+impl MultiValueRange {
+    pub fn check_datatype(&self, datatype: Datatype) -> TileDBResult<()> {
+        check_datatype!(self, datatype);
+        Ok(())
+    }
+}
+
+macro_rules! multi_value_range_try_from {
+    ($($V:ident : $U:ty),+) => {
+        $(
+            impl TryFrom<(u32, Box<[$U]>, Box<[$U]>)> for MultiValueRange {
+                type Error = crate::error::Error;
+                fn try_from(value: (u32, Box<[$U]>, Box<[$U]>)) ->
+                        TileDBResult<MultiValueRange> {
+                    if value.0 < 2 || value.0 == u32::MAX {
```

Comment:
Checking that I follow - `value.0` is meant to be the length?  Is the idea with `value.0 < 2` that you should (must) use a `SingleValueRange` instead?

How come you don't just check `value.1.len()` and `value.2.len()` against 2 and against each other? That sounds more user-friendly.  Is it because the `u32` here holds a fixed cell val num?

Reply:
This would probably be more clear if that were correctly typed as a CellValNum. The logic here is saying "If you have a CellValNum::Fixed(1) or CellValNum::Var, go use the correct range type.

Fun fact, this isn't even used for domains because core doesn't support them, but it felt reasonable to include since it'd be kinda weird to not be able to create them at all.

Reply:
How come you didn't just use `CellValNum`? That would be a lot clearer

Reply:
Totes just saw this in my email so I'm un-resolving this thread. We should probably codify a better protocol for when to resolve an open thread like this rather than relying on timeouts like I did here.

To answer the question though, I basically hadn't internalized the `CellValNum` type at the time I wrote this. I was relying on my knowledge of internals rather than preventing the edge cases. As in, I was more worried about user experience with syntax than forcing correctness by construction.

I'll go back and change this to something better.

Also, it occurs to me that `CellValNum` only has the `Fixed` and `Var` variants. I'm pretty sure it should have `Single` `Multi` and `Var` variants to correctly express the current constraints. That'll probably be a decent change.

Reply:
Reasons favoring `CellValNum::Single`:
- clarifies some pattern matching,`CellValNum::Fixed(nz) if nz.get() == 1` could be changed to `CellValNum::Single`
- construction of `CellValNum::Fixed(1)` is less work

Reasons against:
- Some code doesn't care whether `nz.get() == 1` and this code will get slightly worse
- To prevent overlap between `CellValNum::Single` and `CellValNum::Multi(1)` we would probably need to invent a new type `BoundedInt<2, u32::MAX - 1>` or something. I don't hate that idea and of course there is already a crate for it.

At some point in a branch I added `fn single() -> Self` which addresses the construction problem. So I feel sort of neutrally about this notion.

But regardless I look forward to seeing this changed to `CellValNum`